internal/iam/authorizers.go

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft and contributors.  All rights reserved.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+package iam
+
+import (
+	"log"
+	"net/url"
+	"strings"
+
+	"github.com/Azure-Samples/azure-sdk-for-go-samples/internal/config"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+)
+
+var (
+	armAuthorizer      autorest.Authorizer
+	batchAuthorizer    autorest.Authorizer
+	graphAuthorizer    autorest.Authorizer
+	keyvaultAuthorizer autorest.Authorizer
+)
+
+// OAuthGrantType specifies which grant type to use.
+type OAuthGrantType int
+
+const (
+	// OAuthGrantTypeServicePrincipal for client credentials flow
+	OAuthGrantTypeServicePrincipal OAuthGrantType = iota
+	// OAuthGrantTypeDeviceFlow for device flow
+	OAuthGrantTypeDeviceFlow
+)
+
+// GrantType returns what grant type has been configured.
+func grantType() OAuthGrantType {
+	if config.UseDeviceFlow() {
+		return OAuthGrantTypeDeviceFlow
+	}
+	return OAuthGrantTypeServicePrincipal
+}
+
+// GetResourceManagementAuthorizer gets an OAuthTokenAuthorizer for Azure Resource Manager
+func GetResourceManagementAuthorizer() (autorest.Authorizer, error) {
+	if armAuthorizer != nil {
+		return armAuthorizer, nil
+	}
+
+	var a autorest.Authorizer
+	var err error
+
+	a, err = getAuthorizerForResource(
+		grantType(), config.Environment().ResourceManagerEndpoint)
+
+	if err == nil {
+		// cache
+		armAuthorizer = a
+	} else {
+		// clear cache
+		armAuthorizer = nil
+	}
+	return armAuthorizer, err
+}
+
+// GetBatchAuthorizer gets an OAuthTokenAuthorizer for Azure Batch.
+func GetBatchAuthorizer() (autorest.Authorizer, error) {
+	if batchAuthorizer != nil {
+		return batchAuthorizer, nil
+	}
+
+	var a autorest.Authorizer
+	var err error
+
+	a, err = getAuthorizerForResource(
+		grantType(), config.Environment().BatchManagementEndpoint)
+
+	if err == nil {
+		// cache
+		batchAuthorizer = a
+	} else {
+		// clear cache
+		batchAuthorizer = nil
+	}
+
+	return batchAuthorizer, err
+}
+
+// GetGraphAuthorizer gets an OAuthTokenAuthorizer for graphrbac API.
+func GetGraphAuthorizer() (autorest.Authorizer, error) {
+	if graphAuthorizer != nil {
+		return graphAuthorizer, nil
+	}
+
+	var a autorest.Authorizer
+	var err error
+
+	a, err = getAuthorizerForResource(grantType(), config.Environment().GraphEndpoint)
+
+	if err == nil {
+		// cache
+		graphAuthorizer = a
+	} else {
+		graphAuthorizer = nil
+	}
+
+	return graphAuthorizer, err
+}
+
+// GetKeyvaultAuthorizer gets an OAuthTokenAuthorizer for use with Key Vault
+// keys and secrets. Note that Key Vault *Vaults* are managed by Azure Resource
+// Manager.
+func GetKeyvaultAuthorizer() (autorest.Authorizer, error) {
+	if keyvaultAuthorizer != nil {
+		return keyvaultAuthorizer, nil
+	}
+
+	// BUG: default value for KeyVaultEndpoint is wrong
```

Comment:
Can you find/create an issue to track fixing this?

Reply:
Yes, absolutely plan to, would like to wait till I can test and file a descriptive issue, hence marked as `BUG` for now.